nexus/saga-recovery/src/recovery.rs

```diff
@@ -0,0 +1,705 @@
+    /// running) and update the set of sagas that we believe may be running.
+    ///
+    /// See the big block comment in the saga recovery background task for more
+    /// on how this works and why.
+    pub fn update_started_sagas(
+        &mut self,
+        log: &slog::Logger,
+        sagas_started_rx: &mut mpsc::UnboundedReceiver<SagaId>,
+    ) -> usize {
+        let (new_sagas, disconnected) = read_all_from_channel(sagas_started_rx);
+        if disconnected {
+            warn!(
+                log,
+                "sagas_started_rx disconnected (is Nexus shutting down?)"
+            );
+        }
+
+        let rv = new_sagas.len();
+        let time_observed = Utc::now();
+        for saga_id in new_sagas {
+            info!(log, "observed saga start"; "saga_id" => %saga_id);
+            assert!(self
+                .sagas_started
+                .insert(
+                    saga_id,
+                    SagaStartInfo {
+                        time_observed,
+                        source: SagaStartSource::StartChannel,
+                    }
+                )
+                .is_none());
+        }
+        rv
+    }
+
+    /// Update the at-rest state based on the results of a recovery pass.
+    pub fn update_after_pass(&mut self, plan: &Plan, execution: &Execution) {
+        let time_observed = Utc::now();
+
+        for saga_id in plan.sagas_inferred_done() {
+            assert!(self.sagas_started.remove(&saga_id).is_some());
+        }
+
+        for saga_id in execution.sagas_recovered_successfully() {
+            assert!(self
+                .sagas_started
+                .insert(
+                    saga_id,
+                    SagaStartInfo {
+                        time_observed,
+                        source: SagaStartSource::Recovered,
+                    }
+                )
+                .is_none());
+        }
+
+        self.remove_next = plan.sagas_maybe_done().collect();
+    }
+}
+
+/// Read all message that are currently available on the given channel (without
+/// blocking or waiting)
+///
+/// Returns the list of messages (as a `Vec`) plus a boolean that's true iff the
+/// channel is now disconnected.
+fn read_all_from_channel<T>(
+    rx: &mut mpsc::UnboundedReceiver<T>,
+) -> (Vec<T>, bool) {
+    let mut values = Vec::new();
+    let mut disconnected = false;
+
+    loop {
+        match rx.try_recv() {
+            Ok(value) => {
+                values.push(value);
+            }
+
+            Err(mpsc::error::TryRecvError::Empty) => break,
+            Err(mpsc::error::TryRecvError::Disconnected) => {
+                disconnected = true;
+                break;
+            }
+        }
+    }
+
+    (values, disconnected)
+}
+
+/// Describes what should happen during a particular recovery pass
+///
+/// This is constructed by the saga recovery background task via
+/// [`Plan::new()`].  That function uses `PlanBuilder`
```

Comment:
nit, not a big deal: `PlanBuilder` is a private internal implementation detail, I'm not convinced it needs to be mentioned in the public API docs for this type? when i saw this, i assumed it was also a user-facing API and went looking for it...